packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3614,9 +3643,145 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+      return;
+    }
+    _handleContextMenuOnScroll(notification);
+  }
+
+  Rect _calculateDeviceRect() {
+    final Size screenSize = MediaQuery.sizeOf(context);
+    final ui.FlutterView view = View.of(context);
+    final double obscuredVertical = (view.padding.top + view.padding.bottom + view.viewInsets.bottom) / view.devicePixelRatio;
+    final double obscuredHorizontal = (view.padding.left + view.padding.right) / view.devicePixelRatio;
+    final Size visibleScreenSize = Size(screenSize.width - obscuredHorizontal, screenSize.height - obscuredVertical);
+    return Rect.fromLTWH(view.padding.left / view.devicePixelRatio, view.padding.top / view.devicePixelRatio, visibleScreenSize.width, visibleScreenSize.height);
+  }
+
+  void _handleContextMenuOnScroll(ScrollNotification notification) {
+    if (_webContextMenuEnabled) {
+      return;
+    }
+    if (!_platformSupportsFadeOnScroll) {
+      _selectionOverlay?.updateForScroll();
+      return;
+    }
+    // When the scroll begins and the toolbar is visible, hide it
+    // until scrolling ends.
+    //
+    // The selection and renderEditable need to be visible within the current
+    // viewport for the toolbar to show when scrolling ends. If they are not
+    // then the toolbar is shown when they are scrolled back into view, unless
+    // invalidated by a change in TextEditingValue.
+    if (notification is ScrollStartNotification) {
+      if (_dataWhenToolbarShowScheduled != null) {
+        return;
+      }
+      final bool toolbarIsVisible = _selectionOverlay != null
+                                  && _selectionOverlay!.toolbarIsVisible
+                                  && !_selectionOverlay!.spellCheckToolbarIsVisible;
+      final Rect? selectionBounds = toolbarIsVisible ? renderEditable.getBoxesForSelection(_value.selection)
+                                           .map((TextBox box) => box.toRect())
+                                           .reduce((Rect result, Rect rect) => result.expandToInclude(rect)) : null;
+      _dataWhenToolbarShowScheduled = toolbarIsVisible ? (value: _value, selectionBounds: selectionBounds!) : null;
+      if (_dataWhenToolbarShowScheduled != null) {
+        _selectionOverlay?.hideToolbar();
+      }
+    } else if (notification is ScrollEndNotification) {
+      if (_dataWhenToolbarShowScheduled == null) {
+        return;
+      }
+      if (_dataWhenToolbarShowScheduled!.value != _value) {
+        // Value has changed so we should invalidate any toolbar scheduling.
+        _dataWhenToolbarShowScheduled = null;
+        _disposeScrollNotificationObserver();
+        return;
+      }
+
+      final Rect deviceRect = _calculateDeviceRect();
+      final bool selectionVisibleInEditable = renderEditable.selectionStartInViewport.value || renderEditable.selectionEndInViewport.value;
+      final Rect selectionBounds = MatrixUtils.transformRect(renderEditable.getTransformTo(null), _dataWhenToolbarShowScheduled!.selectionBounds);
```

Comment:
Is this similar to selectionVisibleInEditable? Why's selectionVisibleInEditable not enough?

Reply:
Also why is selectionBounds computed when the scrolling starts instead of when it stops?

Reply:
`selectionVisibleInEditable` is not accurate enough because `renderEditable.selectionStartInViewport` and `renderEditable.selectionEndInViewport` only considers whether the selection is visible in `RenderEditable`s local coordinates. 

`selectionOverlapsWithDeviceRect` account for system overlays like the keyboard and system ui.

`_selectionInViewport` considers the `TextField` being in a `Scrollable`/ nested `Scrollable`s. This accounts for the case when the outer scroll view may occlude the `TextField` selection.

> Also why is selectionBounds computed when the scrolling starts instead of when it stops?

I compute the selection `Rect` during the scroll start, these are in local coordinates of the `RenderEditable` so they should not change on scroll. When the scroll ends I compute the transformed `Rect` for selectionBounds.

Reply:
> these are in local coordinates of the RenderEditable so they should not change on scroll.

What's the advantage of doing this, over computing this on scroll end? You don't have to make this assumption if you do this on scroll end right?

> only considers whether the selection is visible in RenderEditables local coordinates.

Ah I see. Thanks for the explanation!

Reply:
By capturing on scroll start we only call `getBoxesForSelection` once, while if we capture on scroll end we call it every time while the toolbar has been scheduled to be shown.